### PR TITLE
[cli][next][build-utils] Variants

### PR DESCRIPTION
**Pull request by AndyBitz**
Add variants. Updates CLI, `build-utils`, and `next`.

### CODE REVIEW

**Review by dferber90 — DISMISSED** [review]

File: packages/cli/src/commands/build/index.ts [hunk 1]
```diff
@@ -774,3 +778,17 @@ function mergeWildcard(
...
+function mergeFlags(
+  buildResults: Iterable<BuildResult | BuildOutputConfig>
+): BuildResultV2Typical['flags'] {
+  let flags: Flag[] = [];
+
+  for (const result of buildResults) {
+    if ('flags' in result) {
+      flags.push(...(result.flags ?? []));
+    }
+  }
```
Comment (Ethan-Arrowood) [inline packages/cli/src/commands/build/index.ts]: I think `Array.prototype.flatMap` is more appropriate here. I find this code not very readable

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap

File: packages/next/src/utils.ts [hunk 2]
```diff
@@ -3025,3 +3025,35 @@ export function isApiPage(page: string | undefined) {
...
+  const pathVariantsManifest = path.join(
+    entryPath,
+    outputDirectory,
+    'variants-manifest.json'
+  );
+
+  const hasVariantsManifest = await fs
+    .access(pathVariantsManifest)
+    .then(() => true)
+    .catch(() => false);
```
Comment (Ethan-Arrowood) [inline packages/next/src/utils.ts:3050]: Can you use the sync [`fs.existsSync`](https://nodejs.org/api/fs.html#fsexistssyncpath) instead?

File: packages/next/src/server-build.ts [hunk 3]
```diff
@@ -1345,6 +1349,17 @@ export async function serverBuild({
...
+  const flags: Flag[] = [];
+  if (variantsManifest) {
+    Object.entries(variantsManifest).forEach(([key, value]) => {
+      flags.push({
+        key,
+        ...value,
+        metadata: value.metadata ?? {},
+      });
+    });
+  }
```
Comment (Ethan-Arrowood) [inline packages/next/src/server-build.ts]: Use map instead

```suggestion
  let flags: Flag[] = variantsManifest ? Object.entries(variantsManifest).map(([key, value]) => ({
        key,
        ...value,
        metadata: value.metadata ?? {},
    })) : [];
```

**Review by Ethan-Arrowood — CHANGES_REQUESTED** [review]

File: packages/next/src/utils.ts (hunk 2 above)
Comment (AndyBitz) [reply packages/next/src/utils.ts:3050]: I've copied this pattern from the other manifest files, I'm not sure if there's a specific reason for that, but for now I'd like to keep it consistent if possible.

Comment (Ethan-Arrowood) [reply packages/next/src/utils.ts:3050]: okay - this will be slower but I'm okay with consistency.

**Review by dferber90 — DISMISSED** [review]

**Review by dferber90 — APPROVED** [review]

**Review by Ethan-Arrowood — APPROVED** [review]
Are there tests you can include with this PR or will that be a follow up?

**Comment by AndyBitz** [comment]
> Are there tests you can include with this PR or will that be a follow up?

A follow up, eventually.

**Pull request merged**